x/precisebank/keeper/burn.go

```diff
@@ -0,0 +1,170 @@
+package keeper
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/kava-labs/kava/x/precisebank/types"
+)
+
+// BurnCoins burns coins deletes coins from the balance of the module account.
+// It will panic if the module account does not exist or is unauthorized.
+func (k Keeper) BurnCoins(ctx sdk.Context, moduleName string, amt sdk.Coins) error {
+	// Custom protection for x/precisebank, no external module should be able to
+	// affect reserves.
+	if moduleName == types.ModuleName {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "module account %s cannot be burned from", moduleName))
+	}
+
+	// Panic errors are identical to x/bank for consistency.
+	acc := k.ak.GetModuleAccount(ctx, moduleName)
+	if acc == nil {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", moduleName))
+	}
+
+	if !acc.HasPermission(authtypes.Burner) {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "module account %s does not have permissions to burn tokens", moduleName))
+	}
+
+	// Ensure the coins are valid before burning
+	if !amt.IsValid() {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidCoins, amt.String())
+	}
+
+	// Get non-ExtendedCoinDenom coins
+	passthroughCoins := amt
+
+	extendedAmount := amt.AmountOf(types.ExtendedCoinDenom)
+	if extendedAmount.IsPositive() {
+		// Remove ExtendedCoinDenom from the coins as it is managed by x/precisebank
+		removeCoin := sdk.NewCoin(types.ExtendedCoinDenom, extendedAmount)
+		passthroughCoins = amt.Sub(removeCoin)
+	}
+
+	// Coins unmanaged by x/precisebank are passed through to x/bank
+	if !passthroughCoins.Empty() {
+		if err := k.bk.BurnCoins(ctx, moduleName, passthroughCoins); err != nil {
+			return err
+		}
+	}
+
+	// No more processing required if no ExtendedCoinDenom
+	if extendedAmount.IsZero() {
+		return nil
+	}
+
+	return k.burnExtendedCoin(ctx, moduleName, extendedAmount)
+}
+
+// burnExtendedCoin burns the fractional amount of the ExtendedCoinDenom from the module account.
+func (k Keeper) burnExtendedCoin(
```

Comment:
i'm convinced the logic is correct. also, you have great test coverage of the different paths. however, after playing around with it and seeing it in many different formulations, i found that nesting the conditions made the control flow slightly easier to follow.

not sold on it, just though that it was worth sharing. let me know what you think @drklee3 & @nddeluca:
```go
// -------------------------------------------------------------------------
// Stateful operations for burn

// handle borrowing a coin to cover insufficient fractional balance
if newFractionalBalance.IsNegative() {
	// a negative new fractional balance means a coin must be borrowed
	newFractionalBalance = newFractionalBalance.Add(types.ConversionFactor())

	// the borrowed coin can come from one of two place:
	if newRemainder.GTE(types.ConversionFactor()) {
		// 1. optimization: borrow from the remainder if it is overflowing
		newRemainder = newRemainder.Sub(types.ConversionFactor())
		integerBurnAmount = integerBurnAmount.AddRaw(1)
	} else {
		// 2. borrow from the module burning the coins
		// the borrowed coin is put into the module reserves to ensure that
		// the new fractional balance is fully backed
		borrowCoin := sdk.NewCoin(types.IntegerCoinDenom, sdkmath.OneInt())
		if err := k.bk.SendCoinsFromModuleToModule(
			ctx,
			moduleName,
			types.ModuleName, // borrowed coin is put in precisebank reserves
			sdk.NewCoins(borrowCoin),
		); err != nil {
			return k.updateInsufficientFundsError(ctx, moduleAddr, amt, err)
		}
	}

} else if newRemainder.GTE(types.ConversionFactor()) {
	// when an overflowing remainder is not used as the borrowed coin,
	// burn the excess coin from precisebank's reserves
	reserveBurnCoins := sdk.NewCoins(sdk.NewCoin(types.IntegerCoinDenom, sdkmath.OneInt()))
	if err := k.bk.BurnCoins(ctx, types.ModuleName, reserveBurnCoins); err != nil {
		return fmt.Errorf("failed to burn %s for reserve: %w", reserveBurnCoins, err)
	}
	newRemainder = newRemainder.Sub(types.ConversionFactor())
}

// Burn the integer amount
if !integerBurnAmount.IsZero() {
	coin := sdk.NewCoin(types.IntegerCoinDenom, integerBurnAmount)
	if err := k.bk.BurnCoins(ctx, moduleName, sdk.NewCoins(coin)); err != nil {
		return k.updateInsufficientFundsError(ctx, moduleAddr, amt, err)
	}
}

// Assign new fractional balance in x/precisebank
k.SetFractionalBalance(ctx, moduleAddr, newFractionalBalance)
// Update remainder for burned fractional coins
k.SetRemainderAmount(ctx, newRemainder)
```

Reply:
alternatively, the above with no nesting (and some intermediate variables)
```go
	overflowingRemainder := newRemainder.GTE(types.ConversionFactor())
	requiresBorrow := newFractionalBalance.IsNegative()

	// -------------------------------------------------------------------------
	// Stateful operations for burn

	// Case #1: (optimization) required borrow can be covered by overflowing remainder
	if requiresBorrow && overflowingRemainder {
		newFractionalBalance = newFractionalBalance.Add(types.ConversionFactor())
		newRemainder = newRemainder.Sub(types.ConversionFactor())
		integerBurnAmount = integerBurnAmount.AddRaw(1)
	}

	// Case #2: required borrow comes from module burning the coins
	// the borrowed coin is put into the module reserves to ensure that
	// the new fractional balance is fully backed
	if requiresBorrow && !overflowingRemainder {
		newFractionalBalance = newFractionalBalance.Add(types.ConversionFactor())
		borrowCoin := sdk.NewCoin(types.IntegerCoinDenom, sdkmath.OneInt())
		if err := k.bk.SendCoinsFromModuleToModule(
			ctx,
			moduleName,
			types.ModuleName, // borrowed coin is put in precisebank reserves
			sdk.NewCoins(borrowCoin),
		); err != nil {
			return k.updateInsufficientFundsError(ctx, moduleAddr, amt, err)
		}
	}

	// Case #3: overflowing remainder with no borrow
	// when an overflowing remainder is not used as the borrowed coin,
	// burn the excess coin from the reserves
	if !requiresBorrow && overflowingRemainder {
		reserveBurnCoins := sdk.NewCoins(sdk.NewCoin(types.IntegerCoinDenom, sdkmath.OneInt()))
		if err := k.bk.BurnCoins(ctx, types.ModuleName, reserveBurnCoins); err != nil {
			return fmt.Errorf("failed to burn %s for reserve: %w", reserveBurnCoins, err)
		}
		newRemainder = newRemainder.Sub(types.ConversionFactor())
	}

	// Case #4: (no-op) no borrow required, no overflowing remainder

	// Burn the integer amount
	if !integerBurnAmount.IsZero() {
		coin := sdk.NewCoin(types.IntegerCoinDenom, integerBurnAmount)
		if err := k.bk.BurnCoins(ctx, moduleName, sdk.NewCoins(coin)); err != nil {
			return k.updateInsufficientFundsError(ctx, moduleAddr, amt, err)
		}
	}

	// Assign new fractional balance in x/precisebank
	k.SetFractionalBalance(ctx, moduleAddr, newFractionalBalance)
	// Update remainder for burned fractional coins
	k.SetRemainderAmount(ctx, newRemainder)
```

maybe the take away is just that i prefer the duplicating the `newFractionalBalance.Add(types.ConversionFactor())` and moving the remainder burn/adjustment up 🤷

Reply:
the second one is great! I like having all the 4 cases clear - refactored